### ISSUE TITLE
On device inference of traced model failed: "java.lang.IllegalStateException: Expected IValue type 2, actual type 7"

### ISSUE THREAD

**Issue by ruichernob** [pytorch/pytorch#45276]
## 🐛 Bug

Unable to perform inference on traced DistilBert model.

## To Reproduce

1. Trace DistilBert model:
```
import torch
import torchvision
from transformers import DistilBertModel, DistilBertTokenizer

model = myDistilBertModel.from_pretrained('distilbert-base-multilingual-cased', torchscript=True)
tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-multilingual-cased')

dummy_input = tokenizer("dummy input", return_tensors='pt')
traced_model = torch.jit.trace(model, dummy_input)
traced_model.save('traced_dist_bert.pt')
```
2. In Android studio load module and attempt to perform inference:
```
import org.pytorch.IValue;
import org.pytorch.Module;
import org.pytorch.Tensor;

Module module = Module.load(assetFilePath(this, "traced_dist_bert.pt"));

try {
            long[] inp_ids = new long[] {101, 31178,   117, 15127, 17835, 10124, 21610, 10112,   102};
            long[] att_mask = new long[] {1, 1, 1, 1, 1, 1, 1, 1, 1};
            long[] shape = new long[] {1, 9};
            Tensor input_ids = Tensor.fromBlob(inp_ids, shape);
            Tensor attention_mask = Tensor.fromBlob(att_mask, shape);
            IValue inp = IValue.from(input_ids);
            boolean is_tensor = inp.isTensor(); // True
            boolean is_tuple = inp.isTuple(); // False
            Tensor output = module.forward(inp, IValue.from(attention_mask)).toTensor();
            float[] scores = output.getDataAsFloatArray();

            System.out.println(scores);
        }
        catch (Exception e) {
            System.out.println(e.getMessage());
            System.out.println(e.toString());
            finish();
        }
```
3. Alternatively I've tried to re-implement DistBertModel class with only single tensor argument in the whole chain of 'forward()' calls (attention mask consists from 1s, all boolean arguments = False etc).
Surprisingly, with single tensor in forward argument, I get the same error: java.lang.IllegalStateException: Expected IValue type 2, actual type 7
for line
```
Tensor output = module.forward(inp).toTensor();
```

## Expected behavior

forward works as expected

## Environment

```
PyTorch version: 1.7.0.dev20200921+cpu
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: 10.1.243
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.0.dev20200921+cpu
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[conda] Could not collect
```

## Additional context

Sample for simplified DistilBertModel:

```
class MultiHeadSelfAttention(nn.Module):
    def __init__(self, config):
        super().__init__()

        self.n_heads = config.n_heads
        self.dim = config.dim
        self.dropout = nn.Dropout(p=config.attention_dropout)

        assert self.dim % self.n_heads == 0

        self.q_lin = nn.Linear(in_features=config.dim, out_features=config.dim)
        self.k_lin = nn.Linear(in_features=config.dim, out_features=config.dim)
        self.v_lin = nn.Linear(in_features=config.dim, out_features=config.dim)
        self.out_lin = nn.Linear(in_features=config.dim, out_features=config.dim)

        self.pruned_heads = set()

    def prune_heads(self, heads):
        attention_head_size = self.dim // self.n_heads
        if len(heads) == 0:
            return
        heads, index = find_pruneable_heads_and_indices(heads, self.n_heads, attention_head_size, self.pruned_heads)
        # Prune linear layers
        self.q_lin = prune_linear_layer(self.q_lin, index)
        self.k_lin = prune_linear_layer(self.k_lin, index)
        self.v_lin = prune_linear_layer(self.v_lin, index)
        self.out_lin = prune_linear_layer(self.out_lin, index, dim=1)
        # Update hyper params
        self.n_heads = self.n_heads - len(heads)
        self.dim = attention_head_size * self.n_heads
        self.pruned_heads = self.pruned_heads.union(heads)

    def forward(self, query):
        key = query.detach().clone() #b
        value = query.detach().clone() #b

        bs, q_length, dim = query.size()
        k_length = key.size(1)
        # assert dim == self.dim, 'Dimensions do not match: %s input vs %s configured' % (dim, self.dim)
        # assert key.size() == value.size()

        dim_per_head = self.dim // self.n_heads

        mask_reshp = (bs, 1, 1, k_length)
        mask_shape = query.size()[:-1]
        mask = torch.ones(mask_shape, device=query.device)  # (bs, seq_length)

        def shape(x):
            """ separate heads """
            return x.view(bs, -1, self.n_heads, dim_per_head).transpose(1, 2)

        def unshape(x):
            """ group heads """
            return x.transpose(1, 2).contiguous().view(bs, -1, self.n_heads * dim_per_head)

        q = shape(self.q_lin(query))  # (bs, n_heads, q_length, dim_per_head)
        k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
        v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)

        q = q / math.sqrt(dim_per_head)  # (bs, n_heads, q_length, dim_per_head)
        scores = torch.matmul(q, k.transpose(2, 3))  # (bs, n_heads, q_length, k_length)
        mask = (mask == 0).view(mask_reshp).expand_as(scores)  # (bs, n_heads, q_length, k_length)
        scores.masked_fill_(mask, -float("inf"))  # (bs, n_heads, q_length, k_length)

        weights = nn.Softmax(dim=-1)(scores)  # (bs, n_heads, q_length, k_length)
        weights = self.dropout(weights)  # (bs, n_heads, q_length, k_length)

        context = torch.matmul(weights, v)  # (bs, n_heads, q_length, dim_per_head)
        context = unshape(context)  # (bs, q_length, dim)
        context = self.out_lin(context)  # (bs, q_length, dim)

        return (context,)

class TransformerBlock(nn.Module):
    def __init__(self, config):
        super().__init__()

        assert config.dim % config.n_heads == 0

        self.attention = MultiHeadSelfAttention(config)
        self.sa_layer_norm = nn.LayerNorm(normalized_shape=config.dim, eps=1e-12)

        self.ffn = FFN(config)
        self.output_layer_norm = nn.LayerNorm(normalized_shape=config.dim, eps=1e-12)

    def forward(self, x, attn_mask=None):
        # Self-Attention
        sa_output = self.attention(
            query=x
        )

        assert type(sa_output) == tuple
        sa_output = sa_output[0]
        sa_output = self.sa_layer_norm(sa_output + x)  # (bs, seq_length, dim)

        # Feed Forward Network
        ffn_output = self.ffn(sa_output)  # (bs, seq_length, dim)
        ffn_output = self.output_layer_norm(ffn_output + sa_output)  # (bs, seq_length, dim)

        output = (ffn_output,)

        return output

class Transformer(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.n_layers = config.n_layers

        layer = TransformerBlock(config)
        self.layer = nn.ModuleList([copy.deepcopy(layer) for _ in range(config.n_layers)])

    def forward(self, x):
        input_shape = x.size()[:-1]
        attn_mask = torch.ones(input_shape, device=x.device)  # (bs, seq_length)

        hidden_state = x
        for i, layer_module in enumerate(self.layer):
            layer_outputs = layer_module(
                x=hidden_state
            )
            hidden_state = layer_outputs[-1]

            assert len(layer_outputs) == 1

        return tuple(v for v in [hidden_state] if v is not None)

class myDistilBertModel(DistilBertPreTrainedModel):
    def __init__(self, config):
        super().__init__(config)

        self.embeddings = Embeddings(config)  # Embeddings
        self.transformer = Transformer(config)  # Encoder

        self.init_weights()

    def get_input_embeddings(self):
        return self.embeddings.word_embeddings


    def set_input_embeddings(self, new_embeddings):
        self.embeddings.word_embeddings = new_embeddings


    def _prune_heads(self, heads_to_prune):
        """Prunes heads of the model.
        heads_to_prune: dict of {layer_num: list of heads to prune in this layer}
        See base class PreTrainedModel
        """
        for layer, heads in heads_to_prune.items():
            self.transformer.layer[layer].attention.prune_heads(heads)

    def forward(
        self,
        input_ids=None
    ):
        inputs_embeds = self.embeddings(input_ids)  # (bs, seq_length, dim)

        return self.transformer(
            x=inputs_embeds
        )
```

**Comment by ruichernob** [comment]
The failure caused by attempt to perform casting of the model output to tensor.
```
Tensor output = module.forward(inp, IValue.from(attention_mask)).toTensor();
```

The following instruction works well:
```
Tensor output = module.forward(inp, IValue.from(attention_mask))[0].toTensor();
```
